### DISCUSSION TITLE
How to read a property at `[index, string]`?

### DISCUSSION THREAD

**Discussion by johnjelinek** [basilisp-lang/basilisp#1094] [General]
Given the following example:
```python
# importing pandas module 
import pandas as pd 

# reading csv file from url 
data = pd.read_csv("https://media.geeksforgeeks.org/wp-content/uploads/nba.csv") 

# creating position and label variables
position = 2
label = 'Name'
	
# calling .at[] method
output = data.at[position, label]

# display
print(output)

```

How do I call the `.at[position, label]` part?

I tried:
```clojure
(def data (pd/read_csv "https://media.geeksforgeeks.org/wp-content/uploads/nba.csv"))
(-> data .iloc (subvec 0 3))
;;=>             Name            Team  Number Position   Age Height  Weight            College     Salary
;;   0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0              Texas  7730337.0
;;   1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0          Marquette  6796117.0
;;   2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0  Boston University        NaN

(-> data (.at (list 2 "Name")))
  ;;=> Traceback (most recent call last):
  ;;   TypeError: '_AtIndexer' object is not callable 

(.- (.at data) (list 2 "Name"))
  ;;=> 
  ;;     exception: <class 'basilisp.lang.compiler.exception.CompilerException'>
  ;;         phase: :analyzing
  ;;       message: host interop field must be a symbol
  ;;          form: (.- (.at data) (list 2 "Name"))
```

**Comment by johnjelinek** [comment]
Seems like maybe a workaround could be:
```clojure
(->> data .loc (drop 2) (take 1) (map #(.-Name %)) first)
  ;;=> "John Holland"
```

**Comment by ikappaki** [comment]
HI @johnjelinek,

this would be `(aget (.-at data) #py (2 "Name"))`

```clojure
> basilisp repl
basilisp.user=> (import [pandas :as pd]) nil
nil

basilisp.user=> (def data (pd/read_csv "https://media.geeksforgeeks.org/wp-content/uploads/nba.csv"))
                (-> data .iloc (subvec 0 3))
            Name            Team  Number Position   Age Height  Weight            College     Salary
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0              Texas  7730337.0
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0          Marquette  6796117.0
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0  Boston University        NaN

basilisp.user=> (aget (.-at data) #py (2 "Name"))
"John Holland"
```

Please also have a look at https://github.com/ikappaki/basilisp-kernel/blob/main/notebooks/pandas-03-select.ipynb where I replicated some examples  from the pandas getting started guide for the [Basilisp Kernel](https://github.com/ikappaki/basilisp-kernel).


Pandas makes heavy use of overloaded indexing operators, including `slice`'s, that don't map one to one directly to `Basilisp` `aget`/`aset` primitives. It is in my bucket list to discuss this with @chrisrink10 at some point to see if he could think of a way to facilitate this type of indexing at the language level ...

I hope this helps

**Comment by johnjelinek** [comment]
Thanks, is there another way to write the following?
```clojure
#py (2 "Name")
```
I tried:
```clojure
(lisp->py (list 2 "Name"))
;;=> #py [2 "Name"]
```
and that didn't pass into `aget` correctly

**Comment by ikappaki** [comment]
The `#py ()` [reader syntax](https://basilisp.readthedocs.io/en/latest/reader.html#data-readers) is a python tuple.

Another way to write this is to use the `tuple` built-in fn:

```clojure
basilisp.user=> (def t (tuple [2 "Name"]))
#'basilisp.user/t

basilisp.user=> t
#py (2 "Name")

basilisp.user=> (type t)
<class 'tuple'>

basilisp.user=> (aget (.-at data) (tuple [2 "Name"]))
"John Holland"
```

**Comment by FalseProtagonist** [comment]
can also do
(import [operator :as py-op])
(py-op/getitem x (python/tuple [a b]))
also often need (python/slice nil)

**Comment by chrisrink10** [comment]
> Pandas makes heavy use of overloaded indexing operators, including slice's, that don't map one to one directly to Basilisp aget/aset primitives. It is in my bucket list to discuss this with @chrisrink10 at some point to see if he could think of a way to facilitate this type of indexing at the language level ...

Seems to me like a decent use case for a specialized library with macros to handle Pandas/Polars/etc. APIs.
